Added a tutorial for writing an AsyncNode in rclpy (#6460)
ros2/ros2_documentation: source/Tutorials/Intermediate/Writing-An-Async-Node-Python.rst

Writing an async node (Python)
==============================

> **Goal:** Create and run a service and client using ``AsyncNode``, the asyncio-native node API.

Reviewer: Since this seems to be specifically about services and clients, should we change the title to mention it, e.g., "Writing an async service and client (Python)"? This would match the other tutorials:

* Writing a simple publisher and subscriber (Python)
* Writing a simple service and client (Python)
* Writing an action server and client (Python)

Then the title and file name (including the entry in `source/Tutorials/Intermediate.rst`) would need to be updated

Author: Following up to our discussion in DMs - the main subject of this guide is the new AsyncNode API and the asyncio runtime, with service and client serving only as the demo. I added "asyncio" to the title to make that more explicit, the new title is "Writing an async node with asyncio (Python)".

**Tutorial level:** Intermediate

**Time:** 20 minutes

.. contents:: Contents
   :depth: 2
   :local:

Background
----------

``AsyncNode`` is an asyncio-native alternative to the executor based ``Node`` you used in the beginner tutorial.
This lets you ``await`` other asyncio operations from inside a ROS 2 callback without blocking the rest of the node.

Beyond services and clients, ``AsyncNode`` supports timers, subscriptions, and publishers as well.
Actions are not yet supported.
Callbacks can be either ``async def`` coroutines or regular ``def`` functions.
Keep in mind that a sync callback can't ``await`` anything.

Prerequisites
-------------

- You should have completed the :doc:`beginner service and client tutorial <../Beginner-Client-Libraries/Writing-A-Simple-Py-Service-And-Client>`.

> - You should be comfortable with basic ``asyncio`` concepts like ``async def``, ``await`` and ``asyncio.run``.

Reviewer: could we link to some relevant pages from the asyncio docs here? Probably these:

* `asyncio`: https://docs.python.org/3/library/asyncio.html
* `async def`/`await`: https://docs.python.org/3/howto/a-conceptual-overview-of-asyncio.html#a-conceptual-overview-part-1-the-high-level
* `asyncio.run`: https://docs.python.org/3/library/asyncio-runner.html#asyncio.run

You can't (easily) add a link to `formatted` text in RST, which is annoying, though.

Author: Added. Great suggestion!

- ``AsyncNode`` currently lives in ``rclpy.experimental`` and requires Python 3.12 or newer.

Tasks
-----

1 Create a package
^^^^^^^^^^^^^^^^^^

Open a new terminal and :doc:`source your ROS 2 installation <../Beginner-CLI-Tools/Configuring-ROS2-Environment>` so that ``ros2`` commands will work.

Navigate into the ``ros2_ws/src`` directory created in a :ref:`previous tutorial <new-directory>` and create a new package:

.. code-block:: console

  $ ros2 pkg create --build-type ament_python python_async_node --dependencies rclpy example_interfaces

2 Write the service node
^^^^^^^^^^^^^^^^^^^^^^^^

Inside the ``ros2_ws/src/python_async_node/python_async_node`` directory, create a new file called ``async_service.py`` and paste the following code within:

.. code-block:: python

  import asyncio

  import rclpy
  from rclpy.experimental import AsyncNode

>   from example_interfaces.srv import Trigger

Reviewer: nit: for consistency with the client file (and for proper alphabetical order):

```suggestion
  from example_interfaces.srv import Trigger
  import rclpy
  from rclpy.experimental import AsyncNode
```

  class TriggerServer(AsyncNode):

      def __init__(self):
          super().__init__('trigger_server')
          self._service = self.create_service(
              Trigger, 'trigger', self._callback
          )

      async def _callback(self, _request, response):
          self.get_logger().info('Incoming trigger request')
          await self.get_clock().sleep(2.0)
          response.success = True
          response.message = 'Slept for 2.0s'
          return response

  async def _async_main():
      with rclpy.init():
          node = TriggerServer()
          await node.run()

  def main():
      asyncio.run(_async_main())

  if __name__ == '__main__':
      main()

2.1 Examine the code
~~~~~~~~~~~~~~~~~~~~

The first ``import`` statements pull in ``asyncio``, ``rclpy``, the ``AsyncNode`` class from ``rclpy.experimental``, and the ``Trigger`` service type.

.. code-block:: python

  import asyncio

  import rclpy
  from rclpy.experimental import AsyncNode

>   from example_interfaces.srv import Trigger

Reviewer: nit: for consistency with the client file (and for proper alphabetical order):

```suggestion
  from example_interfaces.srv import Trigger
  import rclpy
  from rclpy.experimental import AsyncNode
```

Its constructor initializes the node with the name ``trigger_server`` and creates a service.

.. code-block:: python

  class TriggerServer(AsyncNode):

      def __init__(self):
          super().__init__('trigger_server')
          self._service = self.create_service(
              Trigger, 'trigger', self._callback
          )

The service callback is a coroutine.
Inside it you can ``await`` any async operation — here the node calls ``self.get_clock().sleep(2.0)``, similar to ``asyncio.sleep`` but with ROS sim time support.
While this coroutine is suspended, the node remains responsive to other callbacks.

.. code-block:: python

  async def _callback(self, _request, response):
      self.get_logger().info('Incoming trigger request')
      await self.get_clock().sleep(2.0)
      response.success = True
      response.message = 'Slept for 2.0s'
      return response

The bottom of the file defines an ``async`` helper that initializes ``rclpy``, constructs the node, and runs it:

.. code-block:: python

  async def _async_main():
      with rclpy.init():
          node = TriggerServer()
          await node.run()

> To run an ``AsyncNode``, you need to ``await node.run()`` - the equivalent of ``rclpy.spin(node)``.

Reviewer: for consistency:

```suggestion
To run an ``AsyncNode``, you need to ``await node.run()`` — the equivalent of ``rclpy.spin(node)``.
```

> Finally, ``main`` itself is regular sync function so that ``ros2 run`` can use it as an entry point — ``ros2 run`` invokes the entry point as a regular function, so it cannot be ``async def``.

Reviewer: ```suggestion
Finally, ``main`` itself is a regular sync function so that ``ros2 run`` can use it as an entry point — ``ros2 run`` invokes the entry point as a regular function, so it cannot be ``async def``.
```

.. code-block:: console

  $ colcon build --symlink-install --packages-select python_async_node

The ``--symlink-install`` flag symlinks the Python source files into the install space instead of copying them, so later edits take effect without rebuilding.

Open a new terminal, navigate to ``ros2_ws``, and source the setup files:

.. tabs::

  .. group-tab:: Linux

    .. code-block:: console

      $ source install/setup.bash

  .. group-tab:: macOS

    .. code-block:: console

      $ . install/setup.bash

  .. group-tab:: Windows

    .. code-block:: console

      $ call install/setup.bat

Now run the service node:

.. code-block:: console

  $ ros2 run python_async_node service

The node will wait for the client's request.

Open another terminal and source the setup files from inside ``ros2_ws`` again.
Start the client node:

.. code-block:: console

  $ ros2 run python_async_node client
  [INFO] [trigger_client]: Service responded: success=True, message="Slept for 2.0s"

Return to the terminal where your service node is running.
You will see that it logged the incoming request:

.. code-block:: console

  [INFO] [trigger_server]: Incoming trigger request

> Enter ``Ctrl+C`` in the server terminal to stop the node.

Reviewer: before: <img width="399" height="45" alt="Image" src="https://github.com/user-attachments/assets/93311a4e-7058-4cea-8e50-b5f97ecee85c" />

after:  <img width="399" height="45" alt="Image" src="https://github.com/user-attachments/assets/c892b166-7322-43d2-8f39-ae819e7d0fbf" />

also, I guess "Press" makes more sense than "Enter" maybe?

```suggestion
Press :kbd:`Ctrl-C` in the server terminal to stop the node.
```

Extensions
----------

The basic ``Trigger`` server above uses ``await clock.sleep(2.0)`` as an example for async work.
Two more substantial use cases are shown below — both are straightforward modifications to ``async_service.py``.

> Because the package was built with ``--symlink-install``, edits to the Python file take effect without rebuilding — stop the running service with ``Ctrl+C`` and start it again to pick up the changes.

Reviewer: ```suggestion
Because the package was built with ``--symlink-install``, edits to the Python file take effect without rebuilding — stop the running service with :kbd:`Ctrl-C` and start it again to pick up the changes.
```

Executing a blocking function from a callback
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. note::

   Just like ``asyncio``, ``AsyncNode`` is not thread-safe.
   If you spawn your own thread, do not access ``AsyncNode`` API from it.
   Schedule that work back onto the event loop thread with ``loop.call_soon_threadsafe`` or ``asyncio.run_coroutine_threadsafe``.

Many useful libraries are still written with synchronous APIs.
Calling them directly from an async callback would block the event loop and stall every other callback on the node.
To work around this, ``asyncio.to_thread`` runs the sync function on a worker thread and returns an awaitable, allowing the node to stay responsive.

> Replace the body of ``async_service.py`` with the following to fetch the latest ROS 2 release tag from GitHub:

Reviewer: I think this GitHub API example might come a bit out of nowhere for readers. I would suggest calling out directly that it is used as an _example_ of synchronous work. Something like:

```suggestion
For example, we can make a synchronous HTTP request to fetch the latest ROS 2 release tag from GitHub through ``asyncio.to_thread``.
Replace the body of ``async_service.py`` with the following:
```

.. code-block:: python

  import asyncio
  import json
  import urllib.request

  import rclpy
  from rclpy.experimental import AsyncNode

>   from example_interfaces.srv import Trigger

Reviewer: nit:

```suggestion
  from example_interfaces.srv import Trigger
  import rclpy
  from rclpy.experimental import AsyncNode
```

      async def _callback(self, _request, response):
          tag = await asyncio.to_thread(self._fetch_latest_release)
          response.success = True
          response.message = f'Latest ROS 2 release: {tag}'
          return response

  async def _async_main():
      with rclpy.init():
          node = TriggerServer()
          await node.run()

  def main():
      asyncio.run(_async_main())

  if __name__ == '__main__':
      main()

Restart the service in one terminal, then run the client in another:

.. code-block:: console

  $ ros2 run python_async_node service

.. code-block:: console

  $ ros2 run python_async_node client
  [INFO] [trigger_client]: Service responded: success=True, message="Latest ROS 2 release: <release-tag>"

Calling another service from inside a callback
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Because callbacks are coroutines, they can ``await`` other ROS 2 service calls just like any other async operation.

Replace the body of ``async_service.py`` so that the ``Trigger`` callback delegates to ``demo_nodes_py``'s ``add_two_ints`` service:

.. code-block:: python

  import asyncio

  import rclpy
  from rclpy.experimental import AsyncNode

>   from example_interfaces.srv import AddTwoInts, Trigger

Reviewer: nit:

```suggestion
  from example_interfaces.srv import AddTwoInts, Trigger
  import rclpy
  from rclpy.experimental import AsyncNode
```
